### PR TITLE
deps.windows: Build libdatachannel natively

### PR DESCRIPTION
**Pull request by pkviet**
### Description
These commits allow building of libdatachannel with msvc.

### Motivation and Context
There are two main motivations: 

1. Currently libdatachannel is cross-compiled with mingw on linux.
However recent commits of libdatachannel have introduced a bug related to threading.
(More info: https://github.com/paullouisageneau/libdatachannel/issues/892)
The bug doesn't occur when the dll is compiled on msvc so an msvc build is much preferred.

2. The cross-compiled dll do not support the C++ API of libdatachannel; there are mangling issues.
Using libdatachannel compiled on msvc solves that second issue.
The C++ API is more developed than the C API so it's a welcome change.

Note that mbedtls ps1 script was re-enabled and patched to support dtls.

### How Has This Been Tested?
I ran locally the scripts for mbedtls and libdatachannel.
I then compiled obs master and checked that the dll is loaded and that the whip service and output
are working when streaming to broadcast-box.

### Types of changes
- Tweak (non-breaking change to improve existing functionality) 
- Performance enhancement (non-breaking change which improves efficiency)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: deps.windows/70-libdatachannel.ps1
Comment (RytoEX) [inline deps.windows/70-libdatachannel.ps1:28]: Do we need to check if `$script:Shared` is false here? @PatTheMav this does seem to mimic existing patterns, but what do you think?

**Review by RytoEX — CHANGES_REQUESTED** [review]
Commit message nits:
* mbedtls -> mbedTLS
* msvc -> MSVC
* dtls -> DTLS
* srtp -> SRTP

File: deps.ffmpeg/70-libdatachannel.zsh
Comment (pkviet) [reply deps.ffmpeg/70-libdatachannel.zsh:5]: the commit is actually the last one on master; since there's no tag i used the last released one which is probably wrong then

Comment (RytoEX) [reply deps.ffmpeg/70-libdatachannel.zsh:5]: If you're using commits from master, then you'll have to make up an appropriate version number.

Comment (pkviet) [reply deps.ffmpeg/70-libdatachannel.zsh:5]: asked Paul-Louis from libatachannel to create a tag

File: deps.windows/70-libdatachannel.ps1
Comment (pkviet) [reply deps.windows/70-libdatachannel.ps1:28]: i ripped that again from Pat's ps1 scripts in https://github.com/obsproject/obs-deps/pull/186

File: deps.windows/10-mbedtls.ps1
Comment (pkviet) [reply deps.windows/10-mbedtls.ps1:52]: no idea, i ripped the file from @PatTheMav  https://github.com/obsproject/obs-deps/pull/186
Since he managed to build fully ffmpeg on msvc, i tried to mimic what he did with mbedtls.

File: deps.ffmpeg/70-libdatachannel.zsh
Comment (RytoEX) [reply deps.ffmpeg/70-libdatachannel.zsh]: I meant that you, in the script here, can make up an appropriately descriptive version string if needed. The version string here is descriptive. It's not functional. Something like "v0.19.0-alpha.3-git" might be fine. It is, admittedly, ideal if there's an actual version/tag to refer to.

Comment (pkviet) [reply deps.ffmpeg/70-libdatachannel.zsh]: gotcha but since i've pinged Paul-Louis from libdatachannel, I've already used what he will set, namely 0.19.0-alpha.4 .

File: deps.windows/70-libdatachannel.ps1
Comment (RytoEX) [inline deps.windows/70-libdatachannel.ps1:6]: Is this restricted to x64 for a specific reason?

Comment (pkviet) [reply deps.windows/70-libdatachannel.ps1:6]: thought we did not support x86 anymore on win ? you're thinking arm64 ? i hope not

Comment (pkviet) [reply deps.windows/70-libdatachannel.ps1:6]: oh https://github.com/obsproject/obs-deps/pull/196
😮

Comment (RytoEX) [reply deps.windows/70-libdatachannel.ps1:6]: Dropping x86 comes in #196, as you've noted.

Comment (pkviet) [reply deps.windows/70-libdatachannel.ps1:6]: so what do you recommend ? enable both for x86 & x64 until Pat's PR lands or if it's likely Pat's PR will land before this one, then stick to x64 ?
I guess it doesn't matter much then

Comment (RytoEX) [reply deps.windows/70-libdatachannel.ps1:6]: Simply don't include this line. It will either be added as part of #196, or will need to be added if that is merged first.

File: deps.ffmpeg/70-libdatachannel.zsh
Comment (pkviet) [reply deps.ffmpeg/70-libdatachannel.zsh]: Paul-louis created a tag which is the one i'm using

File: deps.windows/70-libdatachannel.ps1
Comment (pkviet) [reply deps.windows/70-libdatachannel.ps1]: done

**Comment by pkviet** [comment]
> Commit message nits:
> 
> * mbedtls -> mbedTLS
> * msvc -> MSVC
> * dtls -> DTLS
> * srtp -> SRTP

done

**Comment by pkviet** [comment]
**Update**
Note to self: will need to update cmake/Modules/CopyMSVCBins.cmake once this is merged since the dll is renamed datachannel.dll when compiled by msvc (against libdtachannel.dll when cross-compiling)

**Comment by RytoEX** [comment]
> **Update** Note to self: will need to update cmake/Modules/CopyMSVCBins.cmake once this is merged since the dll is renamed datachannel.dll when compiled by msvc (against libdtachannel.dll when cross-compiling)

You can probably do this independently on the obs-studio side and check for both filename patterns.

**Comment by pkviet** [comment]
> > **Update** Note to self: will need to update cmake/Modules/CopyMSVCBins.cmake once this is merged since the dll is renamed datachannel.dll when compiled by msvc (against libdtachannel.dll when cross-compiling)
> 
> You can probably do this independently on the obs-studio side and check for both filename patterns.

Ok will try to do that

File: deps.windows/70-libdatachannel.ps1
Comment (RytoEX) [inline deps.windows/70-libdatachannel.ps1:39]: Use CMake variable type specifiers (`:BOOL`).

**Review by RytoEX — CHANGES_REQUESTED** [review]
I'd prefer if you reverted commit https://github.com/obsproject/obs-deps/commit/a3c9ed2d08e07c9ac3da55cd7dddd61e31f3a4d7, and then apply further changes to `deps.windows/10-mbedtls.ps1.disabled`/`deps.windows/10-mbedtls.ps1` on top of that.

Commit message nit (second commit message subject):
* mbedtls -> mbedTLS

@PatTheMav is this PR an acceptable intermediate step between the current situation and #186 to enable using the latest libdatachannel, or do you have objections/concerns?

File: deps.windows/70-libdatachannel.ps1
Comment (pkviet) [reply deps.windows/70-libdatachannel.ps1]: fixed

**Comment by pkviet** [comment]
> I'd prefer if you reverted commit [a3c9ed2](https://github.com/obsproject/obs-deps/commit/a3c9ed2d08e07c9ac3da55cd7dddd61e31f3a4d7), and then apply further changes to `deps.windows/10-mbedtls.ps1.disabled`/`deps.windows/10-mbedtls.ps1` on top of that.

done

> 
> Commit message nit (second commit message subject):
> 
> * mbedtls -> mbedTLS
> 

done

File: deps.windows/10-mbedtls.ps1
Comment (pkviet) [reply deps.windows/10-mbedtls.ps1:52]: since there's no word from Pat, and since it's his code, i'll resolve for now.

File: deps.windows/70-libdatachannel.ps1
Comment (pkviet) [reply deps.windows/70-libdatachannel.ps1:27]: since there's no word from Pat, and since it's his code, i'll resolve for now.

**Comment by PatTheMav** [comment]
Given that we'll have no choice but to statically link MbedTLS on Windows everywhere (because its source code is currently inherently incompatible with how DLLs work), it is inevitable that `libdatachannel` will carry its own implementation of it.

Can't we use MbedTLS version 3.4.0 here already? It sounds like we end up in a convoluted situation:

* We build MbedTLS 3.2.1 in `deps.windows`, because `libdatachannel` requires it, but there are issues with MbedTLS 3.4.0 last I remember
* Moving forward we would potentially build MbedTLS 3.4.0 in `deps.ffmpeg`
* This will _only work on Windows_ because MbedTLS is individually and statically linked

As long as no other dependency in `deps.windows` also requires MbedTLS this can work, but once we might add another dependency (one that might even need a more recent MbedTLS version) this solution will stop working.

I'm fine with moving forward with it for now, as we can cross that bridge once we get there, we just need to be aware of it.

**Comment by pkviet** [comment]
Yeah we can use mbedTLS 3.4.0. I did test with it and it still worked fine with MSVC libdatachannel.
@RytoEX waiting for Ryan okeying it or not.

**Comment by pkviet** [comment]
I've updated pre-emptively mbedTLS to 3.4.0, pending Ryan's decision. It's the last commit, so it can be easily dropped.
I retested the DLL and it compiles fine and streams fine to a whip endpoint.

**Comment by RytoEX** [comment]
> * but there are issues with MbedTLS 3.4.0 last I remember

IIRC, 3.3.0 has issues, not 3.4.0. I've discussed elsewhere that the designated initializers issue in 3.3.0 was inadvertently fixed in 3.4.0. Are other issues specific to 3.4.0 for us that I'm forgetting?

This PR originally kept mbedTLS at 3.2.1, and updating to 3.4.0 was addressed separately in #185, as both 3.2.1 and 3.4.0 work for our purposes. While I generally prefer that PRs are narrowly scoped (i.e., one to only change compilation methods, one to change versions), I mostly care that it's a separate commit so that reverts are easily done, if needed.

**Review by PatTheMav — APPROVED** [review]

**Review by RytoEX — APPROVED** [review]
Discussed off-thread, and this is fine as-is. Will wait for CI to finish.

**Pull request merged**
